contrib/win32/win32compat/misc.c

```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+	/* initialize connection to local security provider */
+	if (LsaConnectUntrusted(&lsa_handle) != 0) {
+		goto cleanup;
 	}
 
-	int total_user_groups = num_local_groups_read + num_domain_groups_read + num_global_universal_groups_read;
-
-	/* populate the output */
-	user_groups = malloc(total_user_groups * sizeof(*user_groups));	
-
-	populate_user_groups(user_groups, &num_user_groups, num_local_groups_read, total_local_groups, (LPBYTE) local_groups, LOCAL_GROUP);
-	if (user_domain) {
-		populate_user_groups(user_groups, &num_user_groups, num_domain_groups_read, total_domain_groups, (LPBYTE)domain_groups, DOMAIN_GROUP);
-		populate_user_groups(user_groups, &num_user_groups, num_global_universal_groups_read, total_global_universal_groups, (LPBYTE)global_universal_groups, GLOBAL_UNIVERSAL_GROUP);
+	/* initialize kerberos authentication package */
+	ULONG lsa_authpackage = 0;
+	LSA_STRING lsa_packagename = { (USHORT)strlen(MICROSOFT_KERBEROS_NAME_A), (USHORT)strlen(MICROSOFT_KERBEROS_NAME_A) + 1, (PCHAR)MICROSOFT_KERBEROS_NAME_A };
+	if (LsaLookupAuthenticationPackage(lsa_handle, &lsa_packagename, &lsa_authpackage) != 0) {
+		goto cleanup;
 	}
-	
-	for (int i = 0; i < num_user_groups; i++)
-		to_lower_case(user_groups[i]);
-
-	print_user_groups(user, user_groups, num_user_groups);
 
-	cleanup:
-		if(local_groups)
-			NetApiBufferFree(local_groups);
-
-		if(domain_groups)
-			NetApiBufferFree(domain_groups);
-
-		if(global_universal_groups)
-			NetApiBufferFree(global_universal_groups);
+	/* create custom identifier for local token source as openssh */
+	TOKEN_SOURCE token_source;
+	PCHAR token_string_const = (PCHAR) "OpenSSH";
+	LSA_STRING token_string = { (USHORT)strlen(token_string_const), (USHORT)strlen(token_string_const) + 1, token_string_const };
+	strcpy_s(token_source.SourceName, sizeof(token_source.SourceName), token_string_const);
+	if (AllocateLocallyUniqueId(&token_source.SourceIdentifier) == 0) {
+		goto cleanup;
+	}
 
-		if(dc_name_utf16)
-			NetApiBufferFree(dc_name_utf16);
-	
-		if(user_name_utf16)
-			free(user_name_utf16);
-				
-		if(user_name)
-			free(user_name);
-
-		*ngroups = num_user_groups;
-		return user_groups;
-}
+	/* initialize kerberos s4u structure with domain account */
+	USHORT s4u_size = (USHORT)(sizeof(KERB_S4U_LOGON) + wcslen(wuser) * sizeof(wchar_t));
+	s4u_logon = (PKERB_S4U_LOGON)calloc(1, s4u_size);
+	s4u_logon->MessageType = KerbS4ULogon;
+	s4u_logon->Flags = 0x8;
+	s4u_logon->ClientUpn.Length = (USHORT)(wcslen(wuser) * sizeof(wchar_t));
+	s4u_logon->ClientUpn.MaximumLength = s4u_logon->ClientUpn.Length;
+	s4u_logon->ClientUpn.Buffer = memcpy(((PBYTE)s4u_logon) +
+		sizeof(KERB_S4U_LOGON), wuser, s4u_logon->ClientUpn.Length);
+
+	/* attempt to create logon token */
+	PVOID ProfileBuffer;
+	ULONG ProfileBufferLength;
+	LUID LogonId;
+	QUOTA_LIMITS Quotas;
+	NTSTATUS SubStatus;
+	if ((LsaLogonUser(lsa_handle, &token_string, Network, lsa_authpackage, s4u_logon, s4u_size, NULL,
```

Comment:
Yes, I can use that, but I'll probably want to modify to accept a parameter that indicates we don't need a trusted LSA connection for simple group identification.  Question for you or @manojampalam.  It looks like generate_s4u_user_token() uses MsV1_0S4ULogon for local accounts which is not supported per the documentation: https://msdn.microsoft.com/en-us/library/windows/desktop/aa378764(v=vs.85).aspx  Do you guys have commitment from the security folks that this is appropriate and has long term support?

Reply:
Side note: generate_s4u_user_token() has a a bunch of small potential buffer overruns in the memcpy_s commands since we are copying beyond the potential bounds of string literals (e.g. copying 8 bytes out of "sshd" when the memory is points to is only guaranteed to have 5 bytes).  It's unlikely this would access memory we don't own, but I'll fix these if I have to touch these functions.

Reply:
Actually it looks like the support statement may just be ambiguous... I’m not if it’s telling us it’s not supported in Windows XP and earlier or later.

Reply:
Right. It is supposedly telling us that it is not supported Vista and below. I've emailed the owners to clarify it.